apollo-router/src/plugins/response_cache/cache_key.rs

```diff
@@ -0,0 +1,213 @@
+        // - query hash: invalidate the entry for a specific query and operation name
+        // - additional data: separate cache entries depending on info like authorization status
+        let mut key = format!(
+            "version:{RESPONSE_CACHE_VERSION}:subgraph:{subgraph_name}:type:{entity_type}:entity:{hashed_entity_key}:representation:{hashed_representation}:hash:{subgraph_query_hash}:data:{additional_data_hash}"
+        );
+
+        if let Some(private_id) = private_id {
+            let _ = write!(&mut key, ":{private_id}");
+        }
+
+        key
+    }
+}
+
+/// Hash subgraph query
+pub(super) fn hash_query(query_hash: &QueryHash) -> String {
+    let mut digest = blake3::Hasher::new();
+    digest.update(query_hash.as_bytes());
+    digest.update(&[0u8; 1][..]);
+
+    digest.finalize().to_hex().to_string()
+}
+
+pub(super) fn hash_additional_data(
+    body: &graphql::Request,
+    context: &Context,
+    cache_key: &CacheKeyMetadata,
+) -> String {
+    let mut hasher = blake3::Hasher::new();
+
+    let repr_key = ByteString::from(REPRESENTATIONS);
+    body.variables
+        .iter()
+        .sorted_by(|a, b| a.0.cmp(b.0))
+        .filter(|(key, _value)| key != &&repr_key)
+        .for_each(|(k, v)| {
+            hasher.update(serde_json::to_string(k).unwrap().as_bytes());
+            hasher.update(":".as_bytes());
+            match v {
+                serde_json_bytes::Value::Object(obj) => {
+                    hasher.update("{".as_bytes());
+                    hash(&mut hasher, obj);
+                    hasher.update("}".as_bytes());
+                }
+                _ => {
+                    hasher.update(serde_json::to_string(v).unwrap().as_bytes());
+                }
+            }
+        });
+
+    cache_key
+        .serialize(Blake3Serializer::new(&mut hasher))
+        .expect("this serializer doesn't throw any errors; qed");
+
+    if let Ok(Some(cache_data)) = context.get::<&str, Object>(CONTEXT_CACHE_KEY) {
+        if let Some(v) = cache_data.get("all") {
+            v.serialize(Blake3Serializer::new(&mut hasher))
+                .expect("this serializer doesn't throw any errors; qed");
+        }
+        if let Some(v) = body
+            .operation_name
+            .as_ref()
+            .and_then(|op| cache_data.get(op.as_str()))
+        {
+            v.serialize(Blake3Serializer::new(&mut hasher))
+                .expect("this serializer doesn't throw any errors; qed");
+        }
+    }
+
+    hasher.finalize().to_hex().to_string()
+}
+
+// Order-insensitive structural hash of the representation value
+pub(super) fn hash_representation(
+    representation: &serde_json_bytes::Map<ByteString, Value>,
+) -> String {
+    let mut digest = blake3::Hasher::new();
+
+    hash(&mut digest, representation);
+
+    digest.finalize().to_hex().to_string()
+}
+
+fn hash(state: &mut blake3::Hasher, fields: &serde_json_bytes::Map<ByteString, Value>) {
```

Comment:
Nit: I think if you change the type signature of this `hash` function, you can reuse it to handle the `variables` in `hash_additional_data`.
```suggestion
/// Hashes elements of a serde_json_bytes::Value::Object when yielded via `map.iter()`.
fn hash<'a, I>(state: &mut blake3::Hasher, fields: I)
where
    I: Iterator<Item = (&'a ByteString, &'a Value)>,
{
```

So the variables code would become this and eliminate the duplicated code:
```rust
    hash(&mut hasher, body.variables.iter().filter(|(key, _value)| key != &&repr_key));
```